libirods_rule_engine_plugin-elasticsearch.cpp

```diff
@@ -1,913 +1,841 @@
-                        boost::format("failed to index metadata [%s] [%s] [%s] for [%s] code [%d] message [%s]")
-                        % _attribute
-                        % _value
-                        % _unit
-                        % _object_path
-                        % response.status_code
-                        % response.text);
-            }
-        }
-        catch(const std::runtime_error& _e) {
-            rodsLog(
-                LOG_ERROR,
-                "Exception [%s]",
-                _e.what());
-            THROW(
-                SYS_INTERNAL_ERR,
-                _e.what());
-        }
-        catch(const std::exception& _e) {
-            rodsLog(
-                LOG_ERROR,
-                "Exception [%s]",
-                _e.what());
-            THROW(
-                SYS_INTERNAL_ERR,
-                _e.what());
-        }
-
-    } // invoke_purge_event_metadata
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
 
+#include <algorithm>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace
+{
+	namespace beast = boost::beast;
+	namespace http = beast::http;
+
+	using json = nlohmann::json;
+
+	auto send_http_request(const std::string_view _service_url,
+	                       http::verb _verb,
+	                       const std::string_view _target,
+	                       const std::string_view _body = "") -> std::optional<http::response<http::string_body>>
+	{
+		// TODO These host/port checks should happen at plugin startup.
+		// There's no point paying for these on every http request.
+
+		if (_service_url.empty()) {
+			rodsLog(LOG_ERROR, "%s: Empty service URL.", __func__);
+			return std::nullopt;
+		}
+
+		namespace urls = boost::urls;
+
+		urls::result<urls::url_view> result = urls::parse_uri(_service_url);
+		if (!result) {
+			rodsLog(LOG_ERROR, fmt::format("{}: Could not parse service URL [{}].", __func__, _service_url).c_str());
+			return std::nullopt;
+		}
+
+		namespace net = boost::asio;
+		using tcp = net::ip::tcp;
+
+		try {
+			net::io_context ioc;
+
+			tcp::resolver resolver{ioc};
+			beast::tcp_stream stream{ioc};
+
+			const auto results = resolver.resolve(result->host(), result->port());
+			stream.connect(results);
+
+			http::request<http::string_body> req{_verb, _target, 11};
+			req.set(http::field::host, "localhost");
+			req.set(http::field::user_agent, "iRODS Indexing Plugin/4.3.1");
+			req.set(http::field::content_type, "application/json");
+
+			if (!_body.empty()) {
+				req.body() = _body;
+				{
+					std::stringstream ss;
+					ss << req;
+					const auto s = ss.str();
+					if (s.size() > 256) {
+						rodsLog(LOG_NOTICE,
+						        fmt::format("{}: sending request = (truncated) [{} ...]", __func__, s.substr(0, 256))
+						            .c_str());
+					}
+					else {
+						rodsLog(LOG_NOTICE, fmt::format("{}: sending request = [{}]", __func__, s).c_str());
+					}
```

Comment:
256 - I just pulled a random number out of my head.

Happy to hear opinions on this, if any.

Reply:
I think the value is fine, so long as we have some justification for choosing it. Can we contrive one? :)

Do you know the size of a typical request? Is there some maximum imposed by `boost::beast::http`? Is it a *problem* to send truncated requests?

Reply:
The requests can be large and will never be truncated.

This is specifically about logging. Syslog will truncate messages which will result in truncated JSON structures. This guards against that and acts as a way for admins to confirm the requests are being sent.

These should debug messages too.

Reply:
256 is a good enough starting point. We can improve it later if needed.

Resolving.